src/orc2/mod.rs

```diff
@@ -0,0 +1,2150 @@
+                object_buffer.memory_buffer,
+            );
+        }
+    }
+}
+
+#[llvm_versions(12.0..=latest)]
+impl From<RTDyldObjectLinkingLayer> for ObjectLayer<'static> {
+    fn from(rt_dyld_object_linking_layer: RTDyldObjectLinkingLayer) -> Self {
+        unsafe { ObjectLayer::new_borrowed(rt_dyld_object_linking_layer.object_layer) }
+    }
+}
+
+/// Represents an RTDyldObjectLinkingLayer a special [`ObjectLayer`].
+// Does not have a drop implementation as calling LLVMOrcDisposeObjectLayer
+// on an RTDyldObjectLinkingLayer segfaults. Does leak memory!!!
+#[llvm_versions(12.0..=latest)]
+#[derive(Debug)]
+#[repr(transparent)]
+#[must_use]
+pub struct RTDyldObjectLinkingLayer {
+    object_layer: LLVMOrcObjectLayerRef,
+}
+
+#[llvm_versions(12.0..=latest)]
+impl RTDyldObjectLinkingLayer {
+    pub fn get<'a>(&'a self) -> ObjectLayer<'a> {
+        unsafe { ObjectLayer::new_borrowed(self.object_layer) }
+    }
+}
+
+#[llvm_versions(13.0..=latest)]
+#[derive(Debug, Clone)]
+#[repr(transparent)]
+pub struct IRTransformLayer<'jit> {
+    ir_transform_layer: LLVMOrcIRTransformLayerRef,
+    _marker: PhantomData<&'jit ()>,
+}
+
+#[llvm_versions(13.0..=latest)]
+impl<'jit> IRTransformLayer<'jit> {
+    unsafe fn new_borrowed(ir_transform_layer: LLVMOrcIRTransformLayerRef) -> Self {
+        assert!(!ir_transform_layer.is_null());
+        IRTransformLayer {
+            ir_transform_layer,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn emit<'ctx>(
+        &self,
+        materialization_responsibility: MaterializationResponsibility,
+        module: ThreadSafeModule<'ctx>,
+    ) {
+        unsafe {
+            LLVMOrcIRTransformLayerEmit(
+                self.ir_transform_layer,
+                materialization_responsibility.materialization_responsibility,
+                module.thread_safe_module,
+            );
+        }
+        forget(module);
+    }
+
+    pub fn set_transform() {
+        todo!();
+    }
+}
+
+#[llvm_versions(12.0..=latest)]
+#[llvm_versioned_item]
+#[derive(Debug)]
+pub struct MaterializationUnit {
+    materialization_unit: LLVMOrcMaterializationUnitRef,
+}
+
+#[llvm_versions(12.0..=latest)]
+impl MaterializationUnit {
+    unsafe fn new(materialization_unit: LLVMOrcMaterializationUnitRef) -> Self {
+        assert!(!materialization_unit.is_null());
+        MaterializationUnit {
+            materialization_unit,
+        }
+    }
+
+    #[llvm_versions(13.0..=latest)]
+    pub fn create<'jit>(
+        name: &str,
+        mut symbols: SymbolFlagsMapPairs,
+        static_initializer: Option<SymbolStringPoolEntry>,
+        materializer: Box<(dyn Materializer + 'jit)>,
+    ) -> Self {
+        unsafe {
+            MaterializationUnit::new(LLVMOrcCreateCustomMaterializationUnit(
+                to_c_str(name).as_ptr(),
+                transmute::<*mut MaterializationUnitCtx, _>(Box::into_raw(Box::new(materializer))),
```

Comment:
```suggestion
                materializer.as_mut() as *mut _ as *mut MaterializationUnitCtx,
```

probably not needed anymore when this also uses a generic `Materializer` instead